### PR TITLE
Theme switching: Create theme/styling hooks and context components

### PR DESCRIPTION
**Pull request by chrispader**
@grgia

### Fixed Issues

$ https://github.com/Expensify/App/issues/21019
PROPOSAL: GH_LINK_ISSUE(COMMENT)

### Tests

- [ ] Verify that no errors appear in the JS console

1. Use demo code from [this branch](https://github.com/margelo/expensify-app-fork/compare/@chrispader/theme-switching..@chrispader/theme-switching-demo) to test these changes.
2. Start NewDot
3. Check if theme for Sidebaris changing in an interval of 3 seconds

### Offline tests

Not needed.

### QA Steps

No QA

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/20173411/aa4ce713-70a8-489c-b6c0-a27cc2005f18

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/20173411/7e815120-8d37-48aa-b35c-48ea69c7a5aa

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/20173411/8ebf8276-240c-4fae-8b2f-0a0ac3116df0

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/20173411/a1f0ff09-3a96-4bcf-a04c-22af94d746f5

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/20173411/fadd79ab-d787-4164-99f0-e93e2c01fe83

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/20173411/3da46e50-d25e-43ff-b8a6-07b50b17ceb0

</details>

### CODE REVIEW

File: contributingGuides/STYLING.md
Comment (grgia) [inline contributingGuides/STYLING.md:5]: I think we will want to define what a "non-theme-related" style is here as well

File: src/CONST.js
Comment (grgia) [inline src/CONST.js:676]: Same comments, change to `THEME`

File: src/ONYXKEYS.js
Comment (grgia) [inline src/ONYXKEYS.js:226]:
```suggestion
   PREFERRED_THEME: 'preferredTheme',
```

or 
 `  NVP_PREFERRED_THEME: 'preferredTheme', `

File: src/styles/themes/useThemePreference.js
Comment (grgia) [inline src/styles/themes/useThemePreference.js:7]: move to CONST

File: src/styles/themes/useThemePreference.js
Comment (grgia) [inline src/styles/themes/useThemePreference.js:16]: Add a comment here I think

File: src/styles/themes/useThemePreference.js
Comment (grgia) [inline src/styles/themes/useThemePreference.js:25]: `system` should be in CONST

**Review by grgia — COMMENTED** [review]
A few comments. Let's chat and get centered on the variable names before proceeding (`colorTheme` vs `Theme` vs `preferredTheme` etc)

File: src/styles/themes/useThemePreference.js
Comment (grgia) [inline src/styles/themes/useThemePreference.js:22]: should this be ?
```suggestion
                const theme = newTheme || CONST.DEFAULT_THEME;
```

File: src/styles/themes/ThemeProvider.js
Comment (grgia) [inline src/styles/themes/ThemeProvider.js:14]: Was this for QA? Once we're ready to merge, copy into body of description and remove?

**Comment by chrispader** [comment]
@grgia I'm still gonna test, if the `react-native-reanimated` patch is still needed, i think we updated the REA version recently

**Comment by chrispader** [comment]
@Santhosh-Sellavel please ignore this mention, @grgia is gonna review this PR

**Comment by Santhosh-Sellavel** [comment]
@grgia This looks like it might require for C+ Review, let me know if need thanks!

**Review by grgia — COMMENTED** [review]
Let's update test steps to include the demo code diffs in code blocks if possible. (current link is broken)

File: contributingGuides/STYLING.md
Comment (grgia) [inline contributingGuides/STYLING.md:9]: depend on or contain? If contain, let's use that term instead

File: src/styles/ThemeStylesProvider.js
Comment (grgia) [inline src/styles/ThemeStylesProvider.js:18]:
```
            backgroundColor: theme.appBG.value,
```
Can you explain this line

File: src/styles/themes/ThemeProvider.js
Comment (grgia) [inline src/styles/themes/ThemeProvider.js:17]:
```suggestion
    const theme = useMemo(() => (themePreference === CONST.<light> ? /* TODO: replace with light theme */ darkTheme : darkTheme), [themePreference]);
```

File: src/styles/themes/useTheme.js
Comment (grgia) [inline src/styles/themes/useTheme.js:7]: can we use `!theme` here?

File: src/styles/useThemeStyles.js
Comment (grgia) [inline src/styles/useThemeStyles.js:7]: same here, can we use !themeStyles

**Review by grgia — COMMENTED** [review]
A few minor comments!

File: babel.config.js
Comment (grgia) [inline babel.config.js:42]: Also, can you explain these reanimated changes?

File: src/styles/ThemeStylesProvider.js
Comment (chrispader) [reply src/styles/ThemeStylesProvider.js:18]: Sorry, this is a leftover from the REA implementation before.

This is where the dynamic style gets initialized and memoized

File: babel.config.js
Comment (chrispader) [reply babel.config.js:42]: [The docs here](https://docs.swmansion.com/react-native-reanimated/docs/fundamentals/installation/) state, that the `react-native-reanimated/plugin` needs to be listed last

**Comment by grgia** [comment]
@Santhosh-Sellavel feel free to C+ review for this one, I would like to get a second pair of eyes. I am aiming to merge today, so if you don't think you can get to it, let me know and I'll complete the checklist

**Comment by grgia** [comment]
So pumped about this! cc @shawnborton for a little sneak peak of the theme switching in action that @chrispader has been working on:

https://github.com/Expensify/App/assets/38015950/4842298a-177c-48e8-8ddb-07dd4238ec1a

**Comment by shawnborton** [comment]
Amazing! Let me know when you are ready for design feedback!

**Comment by Santhosh-Sellavel** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos

https://github.com/Expensify/App/assets/85645967/3e360dea-b90e-4f47-8b2b-66149a9e39b3

File: src/styles/themes/useThemePreference.js
Comment (chrispader) [reply src/styles/themes/useThemePreference.js]: I added the `PreferredThemeProvider` and context, seems like a cleaner solution after all! 👍

**Comment by chrispader** [comment]
@Santhosh-Sellavel i resolved your change requests :)

**Review by grgia — DISMISSED** [review]

**Comment by shawnborton** [comment]
I noticed some of the colors seem off. Where is the best place to provide that feedback?

**Comment by grgia** [comment]
@shawnborton we haven't added the theme yet, I'll get a PR up for that next week so we can QA the new theme

**Comment by shawnborton** [comment]
Sounds good!

**Review by Santhosh-Sellavel — APPROVED** [review]
LGTM!

**Comment by Santhosh-Sellavel** [comment]
@grgia wait

File: src/styles/themes/useThemePreference.js
Comment (Santhosh-Sellavel) [inline src/styles/themes/useThemePreference.js:4]: Can we delete this

**Review by Santhosh-Sellavel — CHANGES_REQUESTED** [review]
Otherwise Fine, cc: @grgia. Feel free to merge post this change

**Review by Santhosh-Sellavel — APPROVED** [review]
LGTM, all yours @grgia!

**Review by Santhosh-Sellavel — APPROVED** [review]

**Review by grgia — APPROVED** [review]
ALRIGHT let's goooo

**Comment by Santhosh-Sellavel** [comment]
@grgia Assign me issue for tracking

File: src/styles/themes/useThemePreference.js
Comment (c3024) [inline src/styles/themes/useThemePreference.js:7]: Setting a default value for `themePreference` as a state variable and then updating it in `useEffect` caused an unnecessary (and then cancelled) network request here #34669

**Pull request merged**
